[WIP] Add Temporal support to Manager (IPFS file uploads)

DallasC opened the pull request:
This PR:

- adds support to upload a file or directory to IPFS directly from the ENS app
- adds docker support for deploying

This adds an additional icon/button to the `Records Item (src/components/SingleName/ResolverAndRecords/RecordsItem.js )` component that when clicked allows the user to log in to a provider and then upload files to IPFS. Right now the only provider is [Temporal](temporal.cloud) but any provider who supports an [IPFS](https://docs.ipfs.io/reference/api/http/) can easily be added. 

Currently we are having a problem saving the login token to local storage. The problem likely stems from lines 109-110 in `src/components/IPFS/Login.js`. It is getting called but it isn't setting the token. We have tried also setting the token with `JSON.stringify()` but that didn't seem to make difference either.

I can be reached through github or on discord dc#8876

You can see this PR deployed at https://ens-app.temporal.cloud

jefflau commented:
Hi @DallasC,

is the link https://ens-app.temporal.cloud still active? I can't seem to connect to it

jefflau commented on src/components/IPFS/Config.js:
```diff
@@ -0,0 +1,19 @@
+const Temporal = {
```
Change this variable name to all caps `TEMPORAL` as it's a bunch of unchanging constants. And change the file name to all lowercase. Capitalised file names denote React components in our repository

jefflau commented on src/components/IPFS/Dropzone.js:
```diff
@@ -0,0 +1,125 @@
+import React, { Component } from 'react'
```
We don't use Component anymore for new components. Please refactor to functional components and [hooks](https://reactjs.org/docs/hooks-reference.html)

jefflau commented on src/components/IPFS/Upload.js:
```diff
@@ -0,0 +1,184 @@
+  align-items: center;
+`
+
+class Upload extends Component {
```
Refactor to functional component and hooks

jefflau commented on src/components/IPFS/Upload.js:
```diff
@@ -0,0 +1,184 @@
+      <Container>
+        {this.state.uploading ? (
+          <Loader withWrap large />
+        ) : this.state.successfullUploaded ? (
```
supposed to be successfully?

jefflau commented on src/components/SingleName/ResolverAndRecords/RecordsItem.js:
```diff
@@ -133,22 +167,35 @@ const RecordItemEditable = ({
-    setConfirmed
+    setConfirmed,
+    startUploading,
+    startAuthorizing,
```
Don't think it's a good idea to add uploading and authorising to this reducer inside RecordItem as it only applies to Temporal/IPFS. Seems to be getting a bit messy in RecordsItem, so might be worth splitting it into it's own component called ContenthashRecord and have a special useEditable reducer for contenthash with the additional fields. This will stop `loggedIn()` being called for other records like address as well when it initialises another `useEditable`

jefflau commented on src/components/hooks.js:
```diff
@@ -33,13 +39,19 @@ export function useEditable(
   const [state, dispatch] = useReducer(reducer, initialState)
 
   const startEditing = () => dispatch({ type: types.START_EDITING })
+  const startUploading = () => dispatch({ type: types.START_UPLOADING })
```
I think you can create a custom hooks that called `useEditable` and merges it into another reducer with the added special IPFS actions and states.

DallasC commented:
Thanks we'll get started on the changes :smile:

bonedaddy commented:
> Hi @DallasC,
> 
> is the link https://ens-app.temporal.cloud still active? I can't seem to connect to it

Hey we only had this up while we were testing it. If you want it back up let me know and I can recreate the docker container.

jefflau commented:
> > Hi @DallasC,
> > is the link https://ens-app.temporal.cloud still active? I can't seem to connect to it
> 
> Hey we only had this up while we were testing it. If you want it back up let me know and I can recreate the docker container.

That'd be great. I'd like to see it so I can comment on the functionality/design and if need by get our designer @rliebert90 involved if something needs to be redesigned/adjusted

bonedaddy commented:
> > > Hi @DallasC,
> > > is the link https://ens-app.temporal.cloud still active? I can't seem to connect to it
> > 
> > 
> > Hey we only had this up while we were testing it. If you want it back up let me know and I can recreate the docker container.
> 
> That'd be great. I'd like to see it so I can comment on the functionality/design and if need by get our designer @rliebert90 involved if something needs to be redesigned/adjusted

Try https://ens-dev.temporal.cloud, it's on our devnet which is behind a 150mbit internet connection that's used for quite a few dev services, so it might be a bit slow during day time hours.

DallasC commented:
Refactored the new class components into functional components. I tried making a the new Contenthash component but it ended up having a lot of duplicated code so for now i went back and cleaned up `RecordsItem` so it's not as messy and easier to follow. Do you have any suggestions on how to split it into the new componet? I think i might not be understanding the full structure

bonedaddy commented:
> Refactored the new class components into functional components. I tried making a the new Contenthash component but it ended up having a lot of duplicated code so for now i went back and cleaned up `RecordsItem` so it's not as messy and easier to follow. Do you have any suggestions on how to split it into the new componet? I think i might not be understanding the full structure

I've also deployed the current version: https://ens-dev.temporal.cloud/

brantlymillegan commented:
Noting here that I just emailed Derrick Foote with some UI feedback. I'll also mention it in my Telegram channel with the Temporal guys. Looking forward to getting this done!

makoto commented:
Hello, just tried the https://ens-dev.temporal.cloud/

<img width="687" alt="Screenshot 2020-03-10 at 11 14 56" src="https://user-images.githubusercontent.com/2630/76307302-7d151d00-62c0-11ea-8e79-c0c56590e1e3.png">

Maybe Brantly has given UI feedback, but here are my comments.

1. How do I sign up?
2. Probably you should have some sort of toggle button so that I can close the temporal login UI.
3. Probably it should have some mention of temporal as external service.

brantlymillegan commented:
> Hello, just tried the https://ens-dev.temporal.cloud/
> 
> <img alt="Screenshot 2020-03-10 at 11 14 56" width="687" src="https://user-images.githubusercontent.com/2630/76307302-7d151d00-62c0-11ea-8e79-c0c56590e1e3.png">
> 
> Maybe Brantly has given UI feedback, but here are my comments.
> 
> 1. How do I sign up?
> 2. Probably you should have some sort of toggle button so that I can close the temporal login UI.
> 3. Probably it should have some mention of temporal as external service.

Yep these were basically my feedback too, I agree

bonedaddy commented:
hey all the latest sets of changes are available at https://ens-dev.temporal.cloud

brantlymillegan commented:
@bonedaddy I tried it and I see all the updates. When I signed up, it turned my screen blank, and I had to refresh the page to then login with the new username and password.

makoto commented:
I also got the same blank page with this error on console

<img width="695" alt="Screenshot 2020-03-11 at 11 07 32" src="https://user-images.githubusercontent.com/2630/76411024-d8124700-6388-11ea-940f-b3f2ddfa3392.png">

bonedaddy commented:
Thanks for finding that, we've updated and redeployed. There's one thing with regards to the signup/login that I'm wondering if we should change. The way it works is that if `ens-app` is running on a site other then https://app.ens.domains/, it uses the dev environment of any valid service provider registered with the app, but if its on the app.ens.domains site, it uses the production environment of any valid service provider registered with the app.

makoto commented on src/components/IPFS/Config.js:
```diff
@@ -0,0 +1,25 @@
+
+export function getConfig(service) {
+  switch (service) {
+    case (service = 'TEMPORAL'):
```
Shouldn't this be `==`?

DallasC replied:
Yup, fixed in be67c21

makoto commented on src/components/IPFS/Config.js:
```diff
@@ -0,0 +1,25 @@
+    case (service = 'TEMPORAL'):
+      return TEMPORAL
+    default:
+      return TEMPORAL
```
Why are they returning `TEMPORAL` in default as well as service = 'TEMPORAL'?

DallasC replied:
Technically we don't really need this right now but later when adding other services this could be a useful helper function. Got rid of the default in be67c21 but maybe get rid of `service === TEMPORAL` or the whole thing?

makoto commented on src/components/IPFS/Dropzone.js:
```diff
@@ -0,0 +1,119 @@
+`
+
+const Dropzone = props => {
+  const [highlight, setHighlight] = useState(false)
```
`highlight` does not seem to be used

DallasC replied:
Got rid of this when switching to the two button approach mentioned below

makoto commented:
The login page styling seems broken on mobile

<img width="394" alt="Screenshot 2020-03-24 at 08 43 55" src="https://user-images.githubusercontent.com/2630/77405828-a1d2bf80-6dab-11ea-9219-b52b48d68067.png">

makoto commented:
> The way it works is that if `ens-app` is running on a site other then https://app.ens.domains/,

@bonedaddy can you point to me the code which applies the logic? I couldn't find where it's been done.

I personally think the approach is fine. Are there any different behavior or feature set when you use dev site as opposed to the production site (apart from they don't share login info)?

makoto commented:
Cannot see logout button on mobile

<img width="402" alt="Screenshot 2020-03-24 at 09 06 17" src="https://user-images.githubusercontent.com/2630/77407873-c41a0c80-6dae-11ea-870f-c6357349f8e8.png">

makoto commented:
I was able to drap & drop but click to select does not seem working. Neither jpeg nor png clickable.

<img width="1121" alt="Screenshot 2020-03-24 at 09 08 38" src="https://user-images.githubusercontent.com/2630/77408793-26274180-6db0-11ea-92c5-6fe1f02ea4b5.png">

makoto commented on .dockerignore:
```diff
@@ -0,0 +1,2 @@
+node_modules
+build
```
please add new line

bonedaddy replied:
done as of https://github.com/ensdomains/ens-app/pull/581/commits/65467e692ff39838d8687ca3f4a3fd89f05a39e5

makoto commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+FROM node:10.18.0-buster AS build-env
+RUN apt-get install -y git 
+RUN mkdir /build
+RUN cd /build && git clone https://github.com/RTradeLtd/ens-app.git
```
TODO FOR ENS TEAM: This to be changed into `ensdomains` once merged.

makoto replied:
Also, is Docker necessary? I am trying to manually run it locally but getting the following error when trying to run our test dataset

```
~/.../ens/ens-app (pr/581)$yarn preTest
yarn run v1.15.2
$ node -r esm src/testing-utils/preTest.js
(node:70418) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'getId' of undefined
    at setupWeb3 (/Users/makoto/work/ens/ens-app/src/testing-utils/preTest.js:13:40)
    at init (/Users/makoto/work/ens/ens-app/src/testing-utils/preTest.js:38:28)
    at Object.<anonymous> (/Users/makoto/work/ens/ens-app/src/testing-utils/preTest.js:80:1)
    at Generator.next (<anonymous>)
(node:70418) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
✨  Done in 4.84s.

```

This is also failing on our CI build https://travis-ci.org/github/ensdomains/ens-app/builds/666226825

bonedaddy replied:
Yea I can change it and push the change, it was only set to our end point for dev work.

Naw it's not necessary and can be removed, it was just easier to be able to test stuff locally and push it to our dev environment using  docker images.

Is that error coming from the docker container or from your host OS that you built the container on? I'll have Dallas look

makoto commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+COPY docker-files/nginx/ens_app.conf /etc/nginx/conf.d/ens_app.conf
+RUN rm /etc/nginx/conf.d/default.conf
+EXPOSE 80/tcp
+CMD ["nginx", "-g", "daemon off;"]
```
Please add new line

bonedaddy replied:
done as of https://github.com/ensdomains/ens-app/pull/581/commits/65467e692ff39838d8687ca3f4a3fd89f05a39e5

makoto commented on Makefile:
```diff
@@ -0,0 +1,3 @@
+docker-build:
+	docker build $(OPTFLAGS) -t rtradetech/ens-app:latest .
+	docker image push rtradetech/ens-app:latest
```
Please add new line

bonedaddy replied:
done as of https://github.com/ensdomains/ens-app/pull/581/commits/65467e692ff39838d8687ca3f4a3fd89f05a39e5

bonedaddy commented:
> > The way it works is that if `ens-app` is running on a site other then https://app.ens.domains/,
> 
> @bonedaddy can you point to me the code which applies the logic? I couldn't find where it's been done.
> 
> I personally think the approach is fine. Are there any different behavior or feature set when you use dev site as opposed to the production site (apart from they don't share login info)?

@DallasC will have to point you to that, as I just did the docker stuff

DallasC commented:
For the login/signup I can add the media queries so it handles mobile. While I'm doing that is there any changes you guys want in terms of UI? I'm not much of a designer so the look right now was more to show the functionality. 

--------


> I was able to drap & drop but click to select does not seem working. Neither jpeg nor png clickable.
> 
> <img alt="Screenshot 2020-03-24 at 09 08 38" width="1121" src="https://user-images.githubusercontent.com/2630/77408793-26274180-6db0-11ea-92c5-6fe1f02ea4b5.png">

This is unfortunately a limit to how the browsers handle file uploads. Clicking on it you can only have one of the following enabled for the file picker
- Single file: Select a single file to upload
- Directory: Select an entire folder to upload but you can't upload individual files

Right now it is set to directory upload so the individual file uploads won't work if you click on it but you are able to upload a folder. This is useful if you want to upload something like a website to IPFS.

The way around this would potentially be to have two buttons one for file upload and one for directory uploads. A lot of websites go this route like google drive. 

![Screen Shot 2020-03-25 at 1 59 02 AM](https://user-images.githubusercontent.com/2628528/77461874-338b0d00-6e3e-11ea-9587-8f2e0b38de87.png)

Drag and drop should work with both individual files and directories it's only the clicking that has to be one or the other.

brantlymillegan commented:
> The way around this would potentially be to have two buttons one for file upload and one for directory uploads.

Yes, please have two separate buttons then, otherwise it's confusing why you can't upload certain files

DallasC commented:
Got rid of the docker stuff since we aren't going to be using it and updated the UI. Also, merged all the latest changes from the main branch

brantlymillegan commented:
I just tested out the latest version [here](https://ens-dev.temporal.cloud/), and it mostly was very smooth. I gave some feedback in our Telegram chat, but here's the same info so we don't lose track of it:

1) The drop shadow around the buttons "Logout" "Cancel" "Save" are cut off on both desktop and mobile

Example:
<img width="1792" alt="Screen Shot 2020-03-26 at 3 24 00 PM" src="https://user-images.githubusercontent.com/30162762/77703934-38a0a700-6f79-11ea-84bc-4904128452ce.png">

2) After successfully uploading the file and saving the new IPFS hash, I think the button options need to change. "Save" no longer needs to be there. Maybe there needs to be "Upload Something Else" button to let you repeat the uploading process with a new file/folder.

3) So I clicked "Cancel" to make the tool go away (which was not intuitive. I think the button to close should be an X in the corner or maybe say "Close Uploader"). Then I clicked the cloud upload icon, which got me back to the options to upload a folder or file, but I'm confused as to why it still lists the "New IPFS Hash" below that (the one I previously uploaded and that is already in the record)

4) On Coinbase Wallet, when I tap "Folder Upload" or "File Upload", nothing happens.

5) Everything *did* work in the MetaMask Mobile browser and Trust Wallet browser

6) After I upload a file (but before I actually save the record), it says "File Successfully Uploaded!". That could make someone think they're done with the process, when they still need to hit "Save" and send the txn to actually put the IPFS hash in the Content record, so that needs to be made more. Also, the "File Successfully Uploaded!" remains even after I hit save and txn is confirmed, which is again confusing as though nothing happened, when something important did just happen (namely, the Content record was updated, so maybe it should say so)

brantlymillegan commented:
@bonedaddy @DallasC Ok, I've tested the integration on desktop and mobile many ties and, except for the fact that it didn't work (for unknown reasons) on Coinbase Wallet, it passes my approval for functionality and UI. @makoto has also taken a close look at it. I think the only thing we'd need before merging is for @jefflau to take a look.

brantlymillegan commented:
@jefflau ok, I think we're ready for you to take a look! You can test it at: https://ens-dev.temporal.cloud/

There are two places to test it in the Manager: (1) If you are setting a Content record for the first time, in which case there's a button that says "New Upload", and (2) If you already have a Content record but want to set a different one, there's an upload icon.

jefflau commented on src/components/IPFS/Login.js:
```diff
@@ -0,0 +1,237 @@
+  ${mq.large`
+    width: 400px;
+  `}
+  ${mq.xLarge`
```
This seems redundant if we have `.large` before

jefflau commented on src/components/SingleName/ResolverAndRecords/AddRecord.js:
```diff
@@ -78,6 +83,44 @@ const Row = styled('div')`
+  margin-bottom: 5px;
+  width: 100%;
+  ${mq.small`
+  margin-left: 20px;  
```
indent this please

jefflau commented on src/components/IPFS/Dropzone.js:
```diff
@@ -0,0 +1,109 @@
+
+  const directoryListToArray = list => {
+    const array = []
+    for (var i = 0; i < list.length; i++) {
```
change this to let

jefflau commented on src/components/IPFS/Dropzone.js:
```diff
@@ -0,0 +1,109 @@
+
+  const fileListToArray = list => {
+    const array = []
+    for (var i = 0; i < list.length; i++) {
```
let

jefflau commented on src/components/IPFS/Login.js:
```diff
@@ -0,0 +1,237 @@
+  const [client, setClient] = useState(getConfig('TEMPORAL'))
+
+  const signup = (email, username, password) => {
+    var data = new FormData()
```
Change this to const

jefflau commented on src/components/IPFS/Login.js:
```diff
@@ -0,0 +1,237 @@
+    data.append('username', username)
+    data.append('password', password)
+    data.append('email_address', email)
+    var req = new XMLHttpRequest()
```
change this to const. And move it to the top of the function

jefflau commented on src/components/IPFS/Login.js:
```diff
@@ -0,0 +1,237 @@
+  }
+
+  const login = (username, password) => {
+    var data = JSON.stringify({
```
same as above. Change to const

jefflau commented on src/components/IPFS/Login.js:
```diff
@@ -0,0 +1,237 @@
+      password: password.toString()
+    })
+
+    var req = new XMLHttpRequest()
```
change to const

jefflau commented on src/components/IPFS/Upload.js:
```diff
@@ -0,0 +1,132 @@
+  const [upload, setUpload] = useState(false)
+  const [uploadError, setUploadError] = useState(false)
+
+  const sendRequest = newfiles => {
```
This function will keep getting created each render. Please refactor this using [`useCallback`](https://reactjs.org/docs/hooks-reference.html#usecallback) so it doesn't get created every time.
